src/macaron/build_spec_generator/jdk_finder.py

```diff
@@ -0,0 +1,345 @@
+            )
+        except InvalidHTTPResponseError as error:
+            logger.error("Failed why trying to download jar file. Error: %s", error)
+            return None
+        except OSError as os_error:
+            logger.critical("Critical %s", os_error)
+            return None
+
+        return get_jdk_version_from_jar(local_artifact_path)
+
+
+def find_jdk_version_from_remote_maven_repo_cache(
+    group_id: str,
+    artifact_id: str,
+    version: str,
+    asset_name: str,
+    remote_maven_repo_url: str,
+    local_cache_repo: str,
+) -> str | None:
+    """Return the jdk version string from an artifact matching a given GAV from a remote maven layout repository.
+
+    This function cache the downloaded artifact in a maven layout https://maven.apache.org/repository/layout.html
+    under ``local_cache_repo``.
+    We assume that the remote maven layout repository supports downloading a file through a HTTPS URL.
+
+    Parameters
+    ----------
+    group_id: str
+        The group ID part of the GAV coordinate.
+    artifact_id: str
+        The artifact ID part of the GAV coordinate.
+    version: str
+        The version part of the GAV coordinate.
+    asset_name: str
+        The name of artifact to download and extract the jdk version.
+    remote_maven_repo_url: str
+        The URL to the remote maven layout repository.
+    local_cache_repo: str
+        The path to a local directory for caching the downloaded artifact used in JDK version
+        extraction.
+
+    Returns
+    -------
+    str | None
+        The version string extract from the artifact (as is) or None
+        ff there is an error, or if we couldn't find any jdk version.
+    """
+    maven_repository_path = construct_maven_repository_path(
+        group_id=group_id,
+        artifact_id=artifact_id,
+        version=version,
+        asset_name=asset_name,
+    )
+
+    local_artifact_path = os.path.join(
+        local_cache_repo,
+        maven_repository_path,
+    )
+    if os.path.isfile(local_artifact_path):
+        return get_jdk_version_from_jar(local_artifact_path)
+
+    gav_path = os.path.dirname(local_artifact_path)
+    os.makedirs(
+        gav_path,
+        exist_ok=True,
+    )
+
+    artifact_url = join_remote_maven_repo_url(
+        remote_maven_repo_url,
+        maven_repository_path,
+    )
+    logger.debug(
+        "Find JDK version from jar at %s, using cache %s",
+        artifact_url,
+        local_artifact_path,
+    )
+    try:
+        download_file(
+            artifact_url,
+            local_artifact_path,
+        )
+    except InvalidHTTPResponseError as error:
+        logger.error("Failed why trying to download jar file. Error: %s", error)
+        return None
+    except OSError as os_error:
+        logger.critical("Critical %s", os_error)
+        return None
+
+    return get_jdk_version_from_jar(local_artifact_path)
+
+
+def find_jdk_version_from_central_maven_repo(
```

Comment:
A few things for this function and the things it relies on:

- The Java extension enum seems to only be used here, and contains only a single value. Maybe it should just be hardcoded? Or stored not as an enum?
- The caching strategy also seems to be somewhat static; however I see some value of having that option.
- The two `find_jdk...` functions are very similar. I would suggest having a single entry point that diverges based on the caching logic (assuming that's being kept), as it is mostly the download part that differs.